packages/jellyfish-api-core/__tests__/category/account/futureswap.test.ts

```diff
@@ -0,0 +1,744 @@
+    await testing.generate(1)
+
+    {
+      // zero source amount is given
+      const fswap: FutureSwap = {
+        address: tslaAddress,
+        amount: '0@TSLA'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \': Amount out of range\', code: -3, method: futureswap')
+    }
+    {
+      // negative source amount is given
+      const fswap: FutureSwap = {
+        address: tslaAddress,
+        amount: '-1@TSLA'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \': Amount out of range\', code: -3, method: futureswap')
+    }
+    {
+      // invlaid source dtoken
+      const fswap: FutureSwap = {
+        address: tslaAddress,
+        amount: '1@INVALID'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \': Invalid Defi token: INVALID\', code: 0, method: futureswap')
+    }
+    {
+      // non loan source token
+      const fswap: FutureSwap = {
+        address: tslaAddress,
+        amount: '1@BTC'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \'Test DFIP2203Tx execution failed:\nCould not get source loan token 1\', code: -32600, method: futureswap')
+    }
+    {
+      // destination is given when futureswap dtoken to dusd
+      const fswap: FutureSwap = {
+        address: tslaAddress,
+        amount: '1@TSLA',
+        destination: 'DUSD'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \'Test DFIP2203Tx execution failed:\nDestination should not be set when source amount is a dToken\', code: -32600, method: futureswap')
+    }
+    {
+      // INVALID destination is given when futureswap dusd to dtoken
+      const fswap: FutureSwap = {
+        address: tslaAddress,
+        amount: '1@DUSD',
+        destination: 'INVALID'
+      }
+      const promise = testing.rpc.account.futureSwap(fswap)
+      await expect(promise).rejects.toThrow(RpcApiError)
+      await expect(promise).rejects.toThrow('RpcApiError: \'Destination token not found\', code: -5, method: futureswap')
+    }
+  })
+
+  it('should not create futureswap when DFIP2203 is disabled for the dtoken', async () => {
+    const tslaAddress = await testing.generateAddress()
+    await testing.rpc.account.accountToAccount(collateralAddress, { [tslaAddress]: '1@TSLA' })
+    await testing.generate(1)
+
+    // deactivate DFIP2203 for dtoken
+    const key = 'v0/token/' + idTSLA + '/dfip2203'
```

Comment:
I am thinking about why eslint allows this. Hmm. https://eslint.org/docs/rules/prefer-template

Reply:
😛 my ts syntax has improved backward as I can see. probably too much switch with Rust and cpp🤪

Reply:
https://github.com/DeFiCh/jellyfish/issues/1330